src/text/placements/combined.cpp

```diff
@@ -0,0 +1,119 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ *****************************************************************************/
+
+//mapnik
+#include <mapnik/text/placements/combined.hpp>
+#include <mapnik/xml_node.hpp>
+
+#include <mapnik/text/placements/list.hpp>
+#include <mapnik/text/placements/simple.hpp>
+
+namespace mapnik
+{
+
+//text_placements_combined class
+text_placements_combined::text_placements_combined(text_placements_ptr simple_placement, text_placements_ptr list_placement)
+: simple_placement_(simple_placement),
+  list_placement_(list_placement)
+{
+}
+
+text_placement_info_ptr text_placements_combined::get_placement_info(double scale_factor, feature_impl const& feature, attributes const& vars) const
+{
+    text_placement_info_ptr simple_placement_info = simple_placement_->get_placement_info(scale_factor, feature, vars);
+    
+    text_placement_info_ptr list_placement_info = list_placement_->get_placement_info(scale_factor, feature, vars);
+    
+    return std::make_shared<text_placement_info_combined>(this, scale_factor, simple_placement_info, list_placement_info);
+}
+    
+void text_placements_combined::add_expressions(expression_set & output) const
+{
+    simple_placement_->add_expressions(output);
+    list_placement_->add_expressions(output);
+}
+
+text_placements_ptr text_placements_combined::from_xml(xml_node const& xml, fontset_map const& fontsets, bool is_shield)
+{
+    text_placements_ptr simple_placement_ptr = text_placements_simple::from_xml(xml, fontsets, is_shield);
+    text_placements_ptr list_placement_ptr = text_placements_list::from_xml(xml, fontsets, is_shield);
+    if(simple_placement_ptr && list_placement_ptr)
+    {
+        text_placements_ptr ptr = std::make_shared<text_placements_combined>(simple_placement_ptr, list_placement_ptr);
+        ptr->defaults.from_xml(xml, fontsets, is_shield);
+        return ptr;
+    }
+    return text_placements_ptr();
+}
+    
+//text_placement_info_combined class
+    
+text_placement_info_combined::text_placement_info_combined(text_placements_combined const* parent, double scale_factor, text_placement_info_ptr simple_placement_info, text_placement_info_ptr list_placement_info)
+: text_placement_info(parent, scale_factor),
+  simple_placement_info_(simple_placement_info),
+  list_placement_info_(list_placement_info)
+{
+}
+    
+void text_placement_info_combined::reset_state()
+{
+    simple_placement_info_->reset_state();
+    list_placement_info_->reset_state();
+}
+    
+bool text_placement_info_combined::next() const
+{
+    //logic to get the next combined point placement
+    if(simple_placement_info_ && list_placement_info_)
+    {
+        //try the simple placement first
+        if(simple_placement_info_->next())
+        {
+            apply_simple_placement(); 
+        }
+        else if(list_placement_info_->next()) //try the list placement options
```

Comment:
I think you should discard first list placement before first call of this `next()`. The first list placement is [`default`](https://github.com/mapnik/mapnik/blob/dc8253ec2a64dbfdc54fefae40fb002d04f9987c/src/text/placements/list.cpp#L37) placement so you run through default placement twice.